mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/session/infly/PushAction.java

```diff
@@ -163,6 +167,9 @@ public void write(Session session, Message message, int mqttId, int qos, Subscri
             if (subscription.isRetry()) {
                 message.setRetry(message.getRetry() + 1);
                 logger.warn("retryPush:{},{},{}", session.getClientId(), message.getMsgId(), message.getRetry());
+            } else if (subscription.isShare()) {
+                String lmqTopic = MixAll.LMQ_PREFIX + StringUtils.replace(message.getOriginTopic(), "/","%");
+                lmqQueueStore.popAck(lmqTopic, subscription.getSharedName(), message);
```

Comment:
Why do you return ack directly after writing here? Isn’t it a success when the client returns ack?

Reply:
> Why do you return ack directly after writing here? Isn’t it a success when the client returns ack?

I didn't quite get your question, But Now, just like you said, the popAck is performed after the client returns?

Or do you want to ack only when the future.isSuccess()?

Another consideration: return popAck here is to ensure that the message is delivered to the consumer, otherwise the message may be lost if it is ack immediately after receiving the pop message

Reply:
the popAck is performed after the client returns